xnmt/training_task.py

```diff
@@ -213,7 +213,9 @@ def advance_epoch(self):
         self._augment_data_initial()
       else:
         self._augment_data_next_epoch()
-    if self.training_state.epoch_num==0 or self.sample_train_sents:
+    if self.training_state.epoch_num==0 or self.sample_train_sents or \
+      isinstance(self.model.src_reader, input_reader.SubwordSampleTextReader) \
+      or isinstance(self.model.trg_reader, input_reader.SubwordSampleTextReader):
```

Comment:
I'm wondering if we should make this more general, i.e. add a ```is_deterministic()``` or ```needs_reload()``` to input readers and using that instead of checking the type. This would make the code directly ready for future input readers that need reloading between epochs (data augmentation, noisy inputs, ..).

Reply:
Very good idea. I agree.